pkg/flexfec/flexfec_encoder.go

```diff
@@ -0,0 +1,184 @@
+// SPDX-FileCopyrightText: 2023 The Pion community <https://pion.ly>
+// SPDX-License-Identifier: MIT
+
+// Package flexfec implements FlexFEC to recover missing RTP packets due to packet loss.
+// https://datatracker.ietf.org/doc/html/rfc8627
+package flexfec
+
+import (
+	"encoding/binary"
+
+	"github.com/pion/interceptor/pkg/flexfec/util"
+	"github.com/pion/rtp"
+)
+
+const (
+	// BaseRTPHeaderSize represents the minium RTP packet header size in bytes.
+	BaseRTPHeaderSize = 12
+	// BaseFecHeaderSize represents the minium FEC payload's header size including the
+	// required first mask.
+	BaseFecHeaderSize = 12
+)
+
+// FlexEncoder implements the Fec encoding mechanism for the "Flex" variant of FlexFec.
+type FlexEncoder struct {
+	baseSN      uint16
+	payloadType uint8
+	ssrc        uint32
+	coverage    *ProtectionCoverage
+}
+
+// NewFlexEncoder returns a new FlexFecEncer.
+func NewFlexEncoder(baseSN uint16, payloadType uint8, ssrc uint32) *FlexEncoder {
+	return &FlexEncoder{
+		baseSN:      baseSN,
+		payloadType: payloadType,
+		ssrc:        ssrc,
+		coverage:    nil,
+	}
+}
+
+// EncodeFec returns a list of generated RTP packets with FEC payloads that protect the specified mediaPackets.
+// This method does not account for missing RTP packets in the mediaPackets array nor does it account for
+// them being passed out of order.
+func (flex *FlexEncoder) EncodeFec(mediaPackets []rtp.Packet, numFecPackets uint32) []rtp.Packet {
+	// Start by defining which FEC packets cover which media packets
+	if flex.coverage == nil {
+		flex.coverage = NewCoverage(mediaPackets, numFecPackets)
+	} else {
+		flex.coverage.ResetCoverage()
+	}
+
+	if flex.coverage == nil {
+		return nil
+	}
+
+	// Generate FEC payloads
+	fecPackets := make([]rtp.Packet, numFecPackets)
+	for fecPacketIndex := uint32(0); fecPacketIndex < numFecPackets; fecPacketIndex++ {
+		fecPackets[fecPacketIndex] = flex.encodeFlexFecPacket(fecPacketIndex)
+	}
+
+	return fecPackets
+}
+
+func (flex *FlexEncoder) encodeFlexFecPacket(fecPacketIndex uint32) rtp.Packet {
+	mediaPacketsIt := flex.coverage.GetCoveredBy(fecPacketIndex)
+	flexFecHeader := flex.encodeFlexFecHeader(
+		mediaPacketsIt,
+		flex.coverage.ExtractMask1(fecPacketIndex),
+		flex.coverage.ExtractMask2(fecPacketIndex),
+		flex.coverage.ExtractMask3(fecPacketIndex),
+	)
+	flexFecRepairPayload := flex.encodeFlexFecRepairPayload(mediaPacketsIt.Reset())
+
+	return rtp.Packet{
+		Header: rtp.Header{
+			Version:        2,
+			Padding:        false,
+			Extension:      false,
+			Marker:         false,
+			PayloadType:    flex.payloadType,
+			SequenceNumber: flex.baseSN,
+			Timestamp:      54243243,
```

Comment:
Hi @pougetat. Could you please explain this fixed Timestamp value = 54243243? Is it taken from RFC?

Reply:
It's a leftover hardcoded value. I'll get this fixed soon.

Reply:
Okay. No problem 👍  Thanks for the explanation)